# Classification model(s):
## preprocess data and train model(s)

#### Import necessary modules

In [11]:
import os
import numpy as np
import pandas as pd

from sklearn.neighbors import KNeighborsClassifier
from sklearn import metrics

from preprocessing import partition_dataset
#from api_requests import get_rank, get_matches 
#from composite_stats import composite_player_stats
#from combine_and_expand_match_data import combine_match_dfs
#from graphs import rank_order

### Make folder paths for data (and potentially a new subfolder)

In [2]:
def make_new_folder(folder):
    ## Check whether the specified path exists or not
    isExist = os.path.exists(folder)
    
    if not isExist:
        ## Create a new directory because it does not exist 
        os.makedirs(folder)
        print("The new directory is created!")

In [3]:
data_path = f'data/'
plot_path = f'plots/'
agg_stats_path = f'agg_stats/'
model_path = f'model/'

make_new_folder(data_path)
make_new_folder(plot_path)
make_new_folder(agg_stats_path)
make_new_folder(model_path)

## Preprocess data

In [4]:
full_df = pd.read_csv(f'{data_path}cleaned_total_list.csv')

features_list = ['KD', 'HS_perc', 'avg_ability_usage', 'avg_dmg_rec', 
         'avg_spent', 'avg_loadout', 'level', 'avg_assists']

X_train, X_test, y_train, y_test = partition_dataset(full_df, features_list)

# Training the KNN model

In [19]:
#Create KNN Classifier
knn = KNeighborsClassifier(n_neighbors=1001)

#Train the model using the training sets
knn.fit(X_train, y_train)

#Predict the response for test dataset
y_pred = knn.predict(X_test)

# Model Accuracy, how often is the classifier correct?
print("Accuracy:", metrics.accuracy_score(y_test, y_pred))

Accuracy: 0.11177042126448747


Neighbors: Accuracy
3: 0.07776389756402248
5: 0.08049367293589654
7: 0.08295741087746085
25: 0.09465149096629422
51: 0.09986813796932473
101: 0.10311842135702223
1001: 0.11177042126448747

This does not seem to be going well. 
After looking into it, KNN models do not do well with high dimensionality.
I do not know where this boundary is, but since I have 8 features I need to either:
1. Reduce number of features and see how well the KNN deals with that
2. Change to a different model